syweb/webclient/app/components/matrix/matrix-service.js

```diff
@@ -243,118 +273,161 @@ function($http, $window, $timeout, $q) {
+                        continue;
+                    }
+                    if (flows[i].stages.indexOf('m.login.email.identity') > -1) {
+                        isThreePidFlow = true;
+                    }
+                    if (flows[i].stages.indexOf('m.login.recaptcha') === -1) {
+                        needsCaptcha = false;
+                    }
+                    var supported = true;
+                    for (var j = 0; j < flows[i].stages.length; ++j) {
+                        if (knownTypes.indexOf(flows[i].stages[j]) === -1) {
+                            console.log("Unknown type: "+flows[i].stages[j]+": discarding flow");
+                            supported = false;
                         }
                     }
+                    if (!supported) continue;
                     
-                    // looks like we can register fine, go ahead and do it.
-                    console.log("Using flow " + JSON.stringify(flows[flowIndex]));
-                    firstRegType = flows[flowIndex].type;
-                    var sessionId = undefined;
-                    
-                    // generic response processor so it can loop as many times as required
-                    var loginResponseFunc = function(response) {
-                        if (response.data.session) {
-                            sessionId = response.data.session;
+                    if ( (isThreePidFlow && useThreePidFlow) || (!isThreePidFlow && !useThreePidFlow) ) {
+                        flowIndex = i;
+                    }
+                }
+
+                console.log("Using flow " + JSON.stringify(flows[flowIndex]));
+                var remainingStages = [];
+                for (var i = 0; i < flows[flowIndex].stages.length; ++i) {
+                    if (completed.indexOf(flows[flowIndex].stages[i]) === -1) {
+                        remainingStages.push(flows[flowIndex].stages[i]);
+                    } 
+                }
+                console.log("Remaining stages: " + JSON.stringify(remainingStages));
+                var currentStageIndex = 0;
+                var currentStage = remainingStages[currentStageIndex];
+
+                if (currentStage == 'm.login.recaptcha' && !captchaResponse) {
+                    deferred.reject({
+                        data: {
+                            errcode: "M_CAPTCHA_NEEDED",
+                            error: "Home server requires a captcha.",
+                            public_key: error.data.params['m.login.recaptcha'].public_key
                         }
-                        console.log("login response: " + JSON.stringify(response.data));
-                        if (response.data.access_token) {
-                            deferred.resolve(response);
+                    });
+                    return;
+                }
+
+                var loginResponseFunc = function(response) {
+                    if (response.status / 100 == 2) {
```

Comment:
`if (response.status < 200 || response.status >= 300)`` is much clearer, I would prefer that here please.